streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -263,26 +261,23 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
 
             long offset = globalConsumer.position(topicPartition);
             final Long highWatermark = highWatermarks.get(topicPartition);
-            final BatchingStateRestoreCallback stateRestoreAdapter =
-                (BatchingStateRestoreCallback) ((stateRestoreCallback instanceof
-                                                     BatchingStateRestoreCallback)
-                                                ? stateRestoreCallback
-                                                : new WrappedBatchingStateRestoreCallback(stateRestoreCallback));
+            final RecordBatchingStateRestoreCallback stateRestoreAdapter =
+                StateRestoreCallbackAdapter.adapt(stateRestoreCallback);
 
             stateRestoreListener.onRestoreStart(topicPartition, storeName, offset, highWatermark);
             long restoreCount = 0L;
 
             while (offset < highWatermark) {
                 try {
                     final ConsumerRecords<byte[], byte[]> records = globalConsumer.poll(pollTime);
-                    final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
+                    final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>();
```

Comment:
I'm sorry; I didn't follow this. 

Are you suggesting to do `final List<ConsumerRecord<byte[], byte[]>> restoreRecords = records.records(topicPartition)`? 

This would return the right type, but `restoreRecords`  is a filtered list of only records with non-null keys. It doesn't look like `ConsumerRecords#records` applies this filter.

Reply:
I need to be specific with my comments :)

I'm suggesting sth like `records.records(topicPartition).stream().filter().collect()`, the main point is that `ConsumerRecords#iterate` which is used as `final ConsumerRecord<byte[], byte[]> record : records` is a pretty complex operator and is unnecessary here since we know it will only contain one topic-partition.

Reply:
aaah. Now I understand.

Reply:
I've done this, but I stuck with the foreach loop (and also went back to it in the adapter), as this code path is hot, and I'm not sure of the overhead of `stream()`.

Reply:
In that case could we just use the foreach loop after `ConsumerRecords#records` to get the filter list from the returned list? I just felt leveraging on `ConsumerRecords#iterator` is unnecessarily costly.

Reply:
I think this is what I've done in the most recent version of the code: https://github.com/apache/kafka/pull/5710/files#diff-e870eab2dedeee93442ba87dc12609a4R277
```
for (final ConsumerRecord<byte[], byte[]> record : records.records(topicPartition)) {
  if (record.key() != null) {
    restoreRecords.add(record);
  }
}
```
Is that what you're thinking?